Format length as M:SS by default

diego-plan9 opened the pull request:
As a followup to #1689 and #1737, here is a work-in-progress for trying to make the track lengths be displayed as `M:SS` instead of the raw number of seconds (and miliseconds).

I'm assuming the intent is that the change is solely a visual one - database should still store the value as a float. To accomplish this, I have followed your suggestion and set `Item.length` as a custom `DurationType(types.Float)` with `format()` and `parse()` methods that handle the conversion between H:MM and float formats (which has been added to `ui` to avoid some duplication). Additionally, I have added a `DurationQuery(NumericQuery)` in order to be able to perform queries using both formats, ie:
```
$ beet ls length:1:00..2:00
$ beet ls length:60..120
```

I'd welcome some pointers about potential implications that I might have missed (as I'm still not sure if this change might break some functionality that depends on the length formatting), and ideas about integration tests or any other test to ensure everything works as intended. It's been a busy week on my end (and next one will probably be - sorry about that), but I plan to add at least white-box tests (basically, for `library.DurationType`, though I'm tempted to also add them for `DateType`, `PathType` and `MusicalKey`, as they are missing) in the next few days.

diego-plan9 commented on beets/ui/__init__.py:
```diff
@@ -416,6 +416,19 @@ def human_seconds_short(interval):
+    Raises ValueError if the conversion cannot take place due to `string` not
+    being in the right format.
+    """
+    match = re.match('^(\d+):([0-5]\d)$', string)
```
A note about using a regexp: I have decided to go through this route to make it consistent with `human_seconds_short` (ie. format should be `(int)M:(int)SS`, where SS must be < 60). Using the `datetime` facilities would restrict M to be < 60.

It might be desirable to make it a bit more flexible (allow HH:MM:SS, or M:SS.zzzzz, etc), and it should not be too hard, but I've opted not to include those for keeping it as an "inverse" of sorts of `raw_seconds_short`.

sampsyo replied:
Perfect. This seems good for now—and hour-long tracks are certainly an edge case.

diego-plan9 commented:
Also, this change might reveal a small output discrepancy on the `info` plugin output with no formatting flag: when using a library query values are shown formatted (`length` as M:SS, `samplerate` right as `khz`, etc); but when using a filename `length` is still displayed as a raw float string (same for `samplerate`, etc).

This seems to be due to printing the values from the `dict` produced by the emitter (which in the case of the filename, is populated from the `MediaFile` fields).

sampsyo commented on beets/dbcore/query.py:
```diff
@@ -653,6 +653,35 @@ def col_clause(self):
+            return None
+        try:
+            # TODO: tidy up circular import
+            from beets.ui import raw_seconds_short
```
Good point on the circular import—and we mostly try to make dbcore a separate component. Maybe the function should live in this module?

sampsyo commented on beets/library.py:
```diff
@@ -195,6 +195,29 @@ def normalize(self, key):
+
+    def format(self, value):
+        # TODO: decide if documenting format_raw_length
+        if not beets.config['format_raw_length'].get(bool):
```
FWIW, `.get(bool)` is implied when using a configuration view as a condition.

sampsyo commented:
Fantastic! You beat me to it. :smiley: This looks wonderful.

For this feature, white-box unit tests seem like the right approach to me. You've even updated one case that looks more like an integration test, so I think that should be fine.

And take your time! There's clearly no rush, so no need to apologize for real-life responsibilities.

diego-plan9 commented:
* Added some tests for the library-specific field types on `library`
* Added changelog entry
* Moved `raw_seconds_short` to `util`

As for the circular import issue mentioned on https://github.com/sampsyo/beets/pull/1749#discussion_r46771794, I have moved `raw_seconds_short()` to `beets.util` instead of `dbcore` - it felt a bit out of place to place it directly on `dbcore`, and seems like other functions on the module are using `beets.util` anyway. Please let me know if another approach is preferred to meet the architectural goals, and also if perhaps the "inverse" functions (`ui.human_seconds()` and `ui.human_seconds_short()` should be moved there as well (there are roughly about ~8 calls to those on the codebase anyway).

The tests are not too extensive but should cover the basic `format()` and `parse()` functionality of the types - please let me know if it would be desirable to make them more detailed. Sorry about the commit-noise on one of the tests, it was caused by me not being aware that the `DateType` output was local time specific.

sampsyo commented:
This looks perfect, and the tests will be a great addition by themselves. Sorry for the delay, but I'm merging this now!
